packages/amplify-migration-codegen-e2e/src/helpers.ts

```diff
@@ -3,21 +3,164 @@ import {
   deleteProjectDir,
   initJSProjectWithProfile,
   addAuthWithDefault,
-  amplifyPushAuth,
+  amplifyPush,
   getProjectMeta,
   getUserPool,
   npmInstall,
   getNpxPath,
   nspawn as spawn,
+  addS3WithGuestAccess,
+  checkIfBucketExists,
+  addFunction,
+  functionBuild,
+  getFunction,
+  addApiWithoutSchema,
+  updateApiSchema,
+  getAppSyncApi,
+  amplifyPushForce,
+  describeCloudFormationStack,
 } from '@aws-amplify/amplify-e2e-core';
 import * as fs from 'fs-extra';
 import { $TSAny } from '@aws-amplify/amplify-cli-core';
 import path from 'node:path';
 import { CloudControlClient, GetResourceCommand } from '@aws-sdk/client-cloudcontrol';
+import { AppSyncClient, GetDataSourceCommand } from '@aws-sdk/client-appsync';
 import { unset } from 'lodash';
+import { createHash } from 'crypto';
+import { userInfo } from 'os';
 
+type AppId = string;
+type ProjectName = string;
+type BranchName = string;
+type SandboxName = string;
+
+type BackendIdentifier =
+  | {
+      namespace: Readonly<AppId>;
+      name: Readonly<BranchName>;
+      type: Readonly<'branch'>;
+      hash?: Readonly<string>;
+    }
+  | {
+      namespace: Readonly<ProjectName>;
+      name: Readonly<SandboxName>;
+      type: Readonly<'sandbox'>;
+      hash?: Readonly<string>;
+    };
+
+const STACK_NAME_LENGTH_LIMIT = 128;
+const AMPLIFY_PREFIX = 'amplify';
+const HASH_LENGTH = 10;
+const NUM_DASHES = 4;
 const pushTimeoutMS = 1000 * 60 * 20; // 20 minutes;
 
+function toStackName(backendId: BackendIdentifier): string {
+  const hash = getHash(backendId);
+
+  // only take the first 50 chars here to make sure there is room in the stack name for the namespace as well
+  const name = sanitizeChars(backendId.name).slice(0, 50);
+
+  const namespaceMaxLength =
+    STACK_NAME_LENGTH_LIMIT - AMPLIFY_PREFIX.length - backendId.type.length - name.length - NUM_DASHES - HASH_LENGTH;
+
+  const namespace = sanitizeChars(backendId.namespace).slice(0, namespaceMaxLength - 1);
+
+  return ['amplify', namespace, name, backendId.type, hash].join('-');
+}
+
+const getHash = (backendId: BackendIdentifier): string =>
+  backendId.hash ??
+  // md5 would be sufficient here because this hash does not need to be cryptographically secure, but this ensures that we don't get unnecessarily flagged by some security scanner
+  createHash('sha512').update(backendId.namespace).update(backendId.name).digest('hex').slice(0, HASH_LENGTH);
+
```

Comment:
Why are we copying this from backend repo?
Can't we just use this package https://github.com/aws-amplify/amplify-backend/blob/main/packages/platform-core/API.md ?

Reply:
Tried importing the package, but it is supported only for ES6 modules and importing this package in gen1 cli which is commonJS module, gives import errors due to syntax mismatch.

Reply:
:(.

In that case, please move the copied code to separate file (class ?) and add verbose comments about where it came from and why.

Reply:
Addressed. [d310c32](https://github.com/aws-amplify/amplify-cli/pull/13952/commits/d310c32e939e79bfc8ab2ae81be8d17123a4072f)
I have piped the output of npx ampx sandbox and searched for the prefix arn:aws:cloudformation:<region>:<accountid>:stack having stack name.